### DISCUSSION TITLE
Publishing a docker with the image of a GitHub runner after running a workflow

### DISCUSSION THREAD

**Discussion by rturrado** [community/community#35459] [Actions] [answered]
Hi everyone!

I have a GitHub C++ project, and I'm using GitHub Actions to build this project on an `ubuntu-22.04` GitHub runner, and run tests. I would like to create a docker with the image of this runner after running a workflow, and publish it in dockerhub, so anyone could just pull it and have a Linux docker with a checkout of my project, and a build output (i.e. they could just go and see the project's source code, run the different targets, and so on).

I understand I cannot simply publish an image of the GitHub runner, and that I need to run my job in a container, and then publish that container. I have seen that you can tell GitHub Actions to do this:

```
jobs:
  blah:
    runs-on: ubuntu-22.04
    container:
      image: XXX
```

My question is:
- can I specify `ubuntu-22.04` as container image  in order to get an image of the `ubuntu`-22.04 GitHub runner (i.e., an Ubuntu 22.04 OS with all the software that GitHub installs for this type of runner)? Or,
- do I have to create a docker image myself, starting from an Ubuntu 22.04 image (e.g. from dockerhub), and installing the software I need (e.g. gcc, cmake, and so on)?

Thanks!
Roberto.

**Comment by airtower-luna** [comment, marked as answer]
The image name you specify for the job container is treated exactly like `docker pull` would. It can be on Docker Hub or any other registry you have access to.

However, I think you're a little confused about the approach here: 

> I understand I cannot simply publish an image of the GitHub runner, and that I need to run my job in a container, and then publish that container.

You *can't* publish the job container, it's discarded after the job. What you'd need to do is write a Dockerfile that does the build you want. You can then build an image based on that Dockerfile in your job and push it to whichever registry you like.

**Reply by rturrado** [reply]
Thanks for the quick response @airtower-luna !

I'm a bit lost, yes, that's true :)

1) Couldn't I do the publish of the container as a last step of the job, before the job is finished? Like:

```
jobs:
  blah:
    runs-on: ubuntu-22.04
    container:
      image: XXX
    steps:
      - check out project
      - build
      - run tests
      - publish container
```

2) Let's say I write a Dockerfile. First thing you specify there is a base image. Something like:

```
FROM: ubuntu-22.04
...
```

My main question above remains. Is there a docker image containing whatever comes in a GitHub `ubuntu-22.04` runner I can use as a base image?

Thanks!

**Reply by airtower-luna** [reply]
> My main question above remains. Is there a docker image containing whatever comes in a GitHub ubuntu-22.04 runner I can use as a base image?

Nothing official, and with all the things stuffed in there it'd make a really bad base for a Docker image (which should usually contain only the dependencies of the shipped software). There are some [unofficial images that are similar](https://github.com/nektos/act#runners), but again they're really not suited as a base for an image you intend to publish. You'd force users to download a lot of unnecessary stuff. :sweat_smile: 

> Couldn't I do the publish of the container as a last step of the job, before the job is finished?

Not of the job container, no. What you *could* do is have a Dockerfile that copies the results of the external (in the Actions job) build instead of building your project as part of the container build. However I wouldn't advise doing that, because it makes it much harder for anyone else to reproduce.

**Reply by rturrado** [reply]
Thanks again! I'm starting to see some light :)

I guess I was just being extra lazy here trying to clone the runner into a docker (also because everything just works on the runner at the moment). But my original idea was to start from a very basic docker image (like an "empty" ubuntu image), check out my project there, and, during the process, check what software I needed to build it (and document that list of dependencies).

I have a couple of questions yet:

1) Should I have a docker image (e.g. ubuntu plus gcc, cmake, and so on) that I could use as base image in my Dockerfile?
    Or would it be better to use an "empty" ubuntu base image and leave the installation of the required software as steps in the Dockerfile?
    I understand that the first option is faster and it works well if you don't need to change the software to install (in which case you'd probably better off creating a new docker image)?

2) You mention to checkout the project, build it, and run it as steps in the Dockerfile?
    Is that preferable to tell GitHub Actions to build a container from a Dockerfile, and then do the checkout/build/run in the container (I don't actually know if you can tell GitHub to do that)?

BTW, I'm just noticing... all this caching stuff we usually do in a GitHub workflow... it doesn't make much sense or simply can't be used when you do everything on a docker, does it? Like, caching the boost library (that should already go in the docker image).

**Reply by airtower-luna** [reply]
> Should I have a docker image (e.g. ubuntu plus gcc, cmake, and so on) that I could use as base image in my Dockerfile?
> Or would it be better to use an "empty" ubuntu base image and leave the installation of the required software as steps in the Dockerfile?

I generally prefer the latter option, because it's easier to understand. Either way you probably don't want to ship the whole build environment to users, so consider [using a multi-stage build](https://docs.docker.com/build/building/multi-stage/). That way you can build in the first container stage and then copy the artifacts to a second one that'll turn into the finished image. That way you need to install only the runtime dependencies into the image users will pull.

> You mention to checkout the project, build it, and run it as steps in the Dockerfile?

Almost. Let GitHub Actions do the checkout, then run `docker build` and `docker push` (or use actions that do it for you) just like you would locally. :wink:

There are some examples using actions in the documentation: https://docs.github.com/en/actions/publishing-packages/publishing-docker-images

And here's an example of my own that uses plain `docker` commands: https://github.com/airtower-luna/hello-ghcr/blob/c80e6fb835bd6d8aa003c3e78d107aac52382061/.github/workflows/build.yaml

> all this caching stuff we usually do in a GitHub workflow... it doesn't make much sense or simply can't be used when you do everything on a docker, does it? Like, caching the boost library (that should already go in the docker image).

Yeah, those don't go together very well. If you build the image locally you have the Docker layer cache, but the GitHub-provided runner VMs are discarded after the job, so it doesn't help. And the layer cache will probably be too big to put it in the Actions cache. :sweat_smile: I'd probably just install packages from whatever distro you pick for your base image, so you don't have to bother compiling it yourself.

**Reply by rturrado** [reply]
Hi again @airtower-luna!

I am trying to close up this first trip into dockers (and with it, also this thread).
I still have a kind of conceptual **question** (not technical) I'd like to ask you.

## What I have

- `publish-docker.yml`: https://github.com/rturrado/the_modern_cpp_challenge/blob/unixlike-builds/.github/workflows/publish_docker.yml. :
  - Triggered every time a new tag is pushed.
     Since that is supposed to happen seldom, I went for the easy option to checkout and build the repository instead of upload/download artifacts from other workflows.
  - Set up system: update & upgrade the system, install gcc, restore/install boost...
  - Checkout my repository, and build it.
     I'm using `ccache` but I wouldn't need that; apart from the fact that it's not working for this job, I don't know why.
  - Set up QEMU (probably don't need this!) and Docker Buildx.
  - Login to Docker Hub.
  - Extract metadata.
  - Build and push the docker.
    Two tags are pushed: `latest` and whatever tag triggered the execution of the workflow, e.g. `unixlike-builds-v1.2.0`.

- `Dockerfile`: https://github.com/rturrado/the_modern_cpp_challenge/blob/unixlike-builds/Dockerfile
  - Starts from an ubuntu:22.04 image.
  - Updates, upgrades, and installs two packages needed to run my binaries.
  - Copies a few binaries from my GitHub repository into a `/tmcppc` folder.

The size of the image is only 185MB, what I think is quite reasonable.
You can download it very quickly and you are ready to play with the project.
I'm happy with this.

## What I initially though I should have

My initial idea was to provide my project's clients with a docker where they would have:
- a system with all the dependencies installed (e.g. ubuntu:22.04, with gcc 12, CMake, ninja, git, boost...), and
- a checkout of my project, already built in different configurations.

So that they could:
1) start playing with the project, executing the binaries, and
2) be able to change the code and compile the project.

I have created as well two images:
- `tmcppc-base` with a system with all the dependencies installed (1.13GB), and
- `tmcpp-full`, based on the previous image, and with a checkout of my project, already built (4.27GB).

However, and this is the **question** I mentioned above:
- does it make any sense to offer these images to the user? (I understand those image sizes are far from acceptable, at least for a project like this). Or,
- would it simply be better to document all the dependencies and the build process, and let the users download and build the project wherever they want?

Thanks!

**Reply by airtower-luna** [reply]
I think it's unlikely anyone would want to download those development images, but providing their Dockerfiles might be helpful. Users could either build them if they want to work inside a container, or read them to see what dependencies they need to install otherwise. Ideally make sure they're easy to read and commented where the purpose of an instruction isn't immediately obvious. :slightly_smiling_face:

**Reply by rturrado** [reply]
Many thanks @airtower-luna for the answer! Appreciated.

**Comment by rturrado** [comment]
Wow! Many thanks for the ample responses and for your time. I definitely have enough material to get going and you put me in the right direction.

**Comment by ossanna16** [comment]
Thank you for asking a great question @rturrado! I'm glad you found our community 🚀
